### PR TITLE
Upload audio to Bloom Library (BL-4322)

### PR DESCRIPTION
**Pull request by andrew-polk**
This change is [Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/2802)

### CODE REVIEW

**Comment by hatton** [comment]
Feel free to merge after name fix.

File: src/BloomExe/Publish/AudioProcessor.cs [hunk 1]
```diff
@@ -87,6 +87,12 @@ internal static string GetAudioFolderPath(string bookFolderPath)
 			return Path.Combine(bookFolderPath, "audio");
 		}
 
+		internal static bool HasAnyAudioFiles(string bookFolderPath)
+		{
+			var audioFolderPath = GetAudioFolderPath(bookFolderPath);
+			return Directory.Exists(audioFolderPath) && Directory.EnumerateFiles(audioFolderPath).Any();
```
Comment (sujeffreyl) [inline src/BloomExe/Publish/AudioProcessor.cs]: Some of the other functions in this class only care about specific file extensions.  Do you think this function would want to check only for valid file extensions or to accept any file (which theoretically might not be created by us)?

File: src/BloomExe/WebLibraryIntegration/BloomS3Client.cs [hunk 2]
```diff
@@ -241,6 +241,32 @@ public void UploadBook(string storageKeyOfBookFolder, string pathToBloomBookDire
 			DeleteFileSystemInfo(new DirectoryInfo(wrapperPath));
 		}
 
+		private void RemoveUnwantedAudioFiles(string destDirName, bool excludeNarrationAudio)
+		{
+			string audioDir = Path.Combine(destDirName, "audio");
```
Comment (sujeffreyl) [inline src/BloomExe/WebLibraryIntegration/BloomS3Client.cs]: AudioProcessor.GetAudioFolderPath() also does the same thing

File: src/BloomExe/WebLibraryIntegration/BloomS3Client.cs [hunk 3]
```diff
@@ -241,6 +241,32 @@ public void UploadBook(string storageKeyOfBookFolder, string pathToBloomBookDire
...
+					if (Guid.TryParse(match.Groups[2].Value, out dummy))
+					{
+						// Always remove all .wav files with guid file names. They are the original recordings.
+						if (excludeNarrationAudio || audioFile.EndsWith(".wav"))
+							RobustFile.Delete(audioFile);
+					}
+				}
+			}
+
+			// We always leave the music files to get uploaded.
```
Comment (sujeffreyl) [inline src/BloomExe/WebLibraryIntegration/BloomS3Client.cs]: This function doesn't work properly if the name of the background music is itself a GUID.

I could see 2 ways to deal with this:
1) Just explicitly write down this scenario in the comments (probably with a TODO) and say it doesn't work yet.  Also, probably need to re-word "we ALWAYS leave the music files to get uploaded" to "we try to leave all the music files to get uploaded" or something like that.
2) In BookStorageTests.cs, it identifies all the narration audio and all the background music which is actually used.  It wouldn't be too hard to do something similar to identify only the narration audio which is theoretically useful.  I was writing a new function (which isn't checked in yet) in HTMLDom to find the nodes part of it...  And then we just need to select the value of the "id" attribute.
 		public static XmlNodeList SelectAudioSentenceElements(XmlElement element)
		{
			return element.SafeSelectNodes(".//span[contains(@class,'audio-sentence')]");
		}

Comment (sujeffreyl) [reply src/BloomExe/WebLibraryIntegration/BloomS3Client.cs]: Oh, i see you explicitly mentioned that in a different function...

File: src/BloomExe/WebLibraryIntegration/BloomS3Client.cs [hunk 4]
```diff
@@ -434,11 +460,28 @@ public string DownloadFile(string bucketName, string storageKeyOfFile)
...
+			// The way we determine the difference is that narration audio files are guids.
+			// This isn't 100% accurate because, in theory, someone could choose a music file which has a guid file name.
+			// But we are living with that possibility for now.
+			if (MusicApi.MusicFileExtensions.Any(end => objectKey.ToLowerInvariant().EndsWith(end)))
+			{
+				var match = RegexFileNameRegex.Match(objectKey);
+				if (match.Success)
+				{
+					Guid dummy;
+					return Guid.TryParse(match.Groups[2].Value, out dummy);
```
Comment (sujeffreyl) [inline src/BloomExe/WebLibraryIntegration/BloomS3Client.cs]: I bear more bad news :(

If you look at the ID generation code in https://github.com/sujeffreyl/BloomDesktop/blob/master/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 1011, you can see that the majority of GUIDs will actually end up prefixed by an 'i'.  Guid.TryParse will return false in this case, which means we will upload the narration even though it says not to.  We need to do a 2nd round and strip out the leading 'i'

File: src/BloomExe/Book/BookStorage.cs [hunk 5]
```diff
@@ -393,57 +398,42 @@ internal static List<string> GetImagePathsRelativeToBook(XmlElement element)
...
+			//Look up which files could actually be used by the book
+			var usedAudioFileNames = new HashSet<string>();
+
+			var backgroundMusicFileNames = GetBackgroundMusicFileNamesReferencedInBook();
+			usedAudioFileNames.AddRange(backgroundMusicFileNames);
+
+			var narrationAudioFileNames = GetNarrationAudioFileNamesReferencedInBook(true);
+			usedAudioFileNames.AddRange(narrationAudioFileNames);
+
+			foreach (var fileName in usedAudioFileNames)
```
Comment (sujeffreyl) [inline src/BloomExe/Book/BookStorage.cs]: now that we refactored out the part that figured out which file names we cared about, we can simplify this and don't need to manually intersect the lists anymore.  Could replace the loop with:

audioFilesToDeleteIfNotUsed.ExceptWith(usedAudioFileNames)

**Pull request merged**